WebXRManager: Fix input sources being sent to wrong controllers (touch AR)

marwie opened the pull request:
Related issue: https://github.com/mrdoob/three.js/issues/23187

**Description**

Adds array of input sources to keep track of controllers being still/currently used on inputSourcesChanged event.
This prevents mixup/re-ordering of touches in AR.

<!-- Remove the line below if is not relevant -->

This contribution is funded by 🌵 [needle](https://needle.tools).

mrdoob commented on src/renderers/webxr/WebXRManager.js:
```diff
@@ -662,13 +686,11 @@ class WebXRManager extends EventDispatcher {
 
-			const inputSources = session.inputSources;
-
 			for ( let i = 0; i < controllers.length; i ++ ) {
```
Wouldn't it be better to just do:

```js
for ( let i = 0; i < inputSources.length; i ++ ) {
```

marwie replied:
This is supposed to guarantue that the controller.update event is called with the input source that was previously associated with it (in the onInputSources changed events where we call onConnect and onDisconnect). And session.inputSources does not necessarily be in the same order if I'm not mistaken (e.g. when using multitouch)

mrdoob replied:
Hmm, what't the purpose of `if ( i >= inputSources.length ) continue;` then?

marwie replied:
Input sources are added to this array when connected to controllers for the first time. Ah ok so you're suggesting to use that to loop through, yeah that makes sense :)

https://github.com/mrdoob/three.js/blob/3414e135c1e520cfb01a43ab251fcc6c6147af37/src/renderers/webxr/WebXRManager.js#L392

mrdoob replied:
Seems like the code is a bit hard to follow then...

marwie replied:
Ok! Do you think it would be easier if the inputSource array was always the same length as controllers?
Or should I better use the inputSourceMap (I was wondering why a map was used here, do you know if there's any specific reason for that? Having two arrays with a direct mapping seems easier to me as there would probably never be that many controllers anyways)

marwie replied:
Hey, any news for this? :)

mrdoob replied:
Sorry, the last few months have been a bit hectic... 😅

marwie replied:
No problem, thanks for reviewing now!

hybridherbst commented:
Might be superseded by
- [23830](https://github.com/mrdoob/three.js/pull/23830),
but I still have to verify.

mrdoob commented:
> but I still have to verify

Let us know when you do 🙏

marwie committed:
fix controller input sources: stable controller index

remove inputSourcesMap and instead use controllerInputSources array

fix linting errors

hybridherbst commented:
Finally got hold of another Hololens 2, here's some findings:

- https://github.com/mrdoob/three.js/pull/23830 doesn't fix multitouch AR cases as it only deals with handedness; we'll rebase and update this PR here to fix that

- there's occassional TypeError: cannot read properties of null (reading 'assetUrl') in XRControllerModelFactory.js:276, not sure where that's suddenly coming from though

- there is a regression in WebXR pointer handling, when using Hololens 2 every time a hand leaves/enters the view a new VisualResponse (sphere + pointer) is spawned and stuck in the air. This wasn't the case ~r137. This affects all of the hands samples; for example here: https://threejs.org/examples/?q=hand#webxr_vr_handinput_pointerdrag. This doesn't affect Quest hands, only Hololens hands

- there's another regression (potentially, not sure if this was the case before already) in hand handling on both Quest and Hololens - when wildly clicking around on boxes in this demo https://threejs.org/examples/?q=hand#webxr_vr_handinput_pointerdrag, I can easily get the pointer click to be "stuck" (hand not pinching anymore but cube is following pointer).

cc @cabanier @mrdoob

hybridherbst commented:
Opened one of the regression issues as
- https://github.com/mrdoob/three.js/issues/24071

hybridherbst commented:
@mrdoob @cabanier any input here? Would be great to have multi-controller AR (multitouch, multiple controllers without handedness) working again.

mrdoob commented on src/renderers/webxr/WebXRManager.js:
```diff
@@ -140,17 +148,15 @@ class WebXRManager extends EventDispatcher {
+				const inputSource = controllerInputSources[ i ];
 
-				}
+				if ( ! inputSource ) continue;
```
```js
if ( inputSource === null ) continue;
```

mrdoob commented on src/renderers/webxr/WebXRManager.js:
```diff
@@ -378,7 +373,38 @@ class WebXRManager extends EventDispatcher {
+
+				let controllerIndex = controllerInputSources.indexOf( inputSource );
+
+				if ( controllerIndex < 0 ) {
```
I think this is what we're checking, right?

```js
if ( controllerIndex === -1 ) {
```

marwie replied:
Yes

mrdoob commented on src/renderers/webxr/WebXRManager.js:
```diff
@@ -378,7 +373,38 @@ class WebXRManager extends EventDispatcher {
+
+					// If all controllers do currently receive input we ignore new ones
+
+					if ( controllerIndex < 0 ) break;
```
Same here?

```js
if ( controllerInded === -1 ) break;
```

marwie committed:
apply review changes

mrdoob commented:
Thanks!
